rfd/0xxx-tls-routing-behind-lb.md

```diff
@@ -0,0 +1,319 @@
+
+Once the connection upgrade support to all protocols are implemented, users can be recommended to upgrade their Teleport
+clusters to single port mode if the cluster is currently serving separate ports.
+
+All scenarios discussed in this section assumes Teleport sits behind a load balancer so connection upgrade for TLS
+Routing is required.
+
+#### 1 - Database Access UX
+
+Common flows already use `tsh` local proxy (since v8) so they will automatically perform the connection upgrade:
+
+- `tsh db connect` (single port mode)
+- `tsh proxy db`
+
+So no UX change to these.
+
+In single port mode, using native clients to connect to Teleport Proxy directly with `tsh db env/config` will NOT work
+through load balancers. Users can use one of the above methods instead.
+
+#### 2 - Application Access UX
+
+Application Access through Teleport Webapp is NOT affected by the issue of interest of this RFD.
+
+On the `tsh` side, TCP, AWS, Azure, and GCP apps always use a local proxy so they will automatically perform the
+connection upgrade when needed. No UX change to these.
+
+However, HTTP apps MUST use `tsh proxy app`.
+
+#### 3 - Kubernetes Access UX
+
+When upgrade is required, using native Kubernetes clients after `tsh kube login` will NOT work as the native clients
+cannot upgrade the connection. `tsh kube login` should advertise users to run `tsh kubectl` and `tsh proxy kube` instead
+of `kubectl version`. `tsh kube credentials` should also error out and suggest using `tsh kubectl` and `tsh proxy kube`.
+
+`tsh kubectl` will automatically starts a local proxy and performs the connection upgrade when needed.
+
+A new `tsh proxy kube` command will be added to support other Kubernetes clients. The command will start a local proxy
+and provide a config for Kubernetes clients to use:
+
+```
+$ tsh proxy kube cluster1 cluster2 -p 8888
+
+Started local Kubernetes server at https://localhost:8888.
+
+Please use the following KUBECONFIG for your Kubernetes applications:
+export KUBECONFIG=<path to kubeconfig>
+```
+
+`tsh proxy kube` should support all flags supported by `tsh kube login` like cluster names, `--as` etc. and can be run
+independently WITHOUT running `tsh kube login` first. If `tsh kube login` has been run perviously, `tsh proxy kube`
+should inherit the flags from the login if they are not specfied during `tsh proxy kube`.
+
+A `--exec` flag will be provided to `tsh proxy kube` to execute a command backed by the local proxy:
+```
+$ tsh proxy kube --exec -- helm install my-chart
+```
+This avoids needing two terminal sessions to run a Kubernetes application.
+
+In addition, we can provide tips to users to improve their UX by utilizing `alias`:
+```
+alias kubectl=`tsh kubectl`
+alias helm=`tsh proxy kube --exec -- helm`
+```
+
+or [`tsh` aliases](https://github.com/gravitational/teleport/blob/master/rfd/0061-tsh-aliases.md):
+```
+aliases:
+    "helm": "$TSH proxy kube --exec helm"
+```
+
+#### 4 - Server Access UX
+
+No UX change.
```

Comment:
"No UX change" as in "tsh ssh will continue to just work even through ALB"? What about connecting with vanilla OpenSSH clients?

Reply:
👍 i missed that. This is `tsh config` right? I think we have to always use `ProxyCommand`. I will look into this and update.

Reply:
The ssh config generated is already calling `tsh proxy ssh` for openssh client:
https://github.com/gravitational/teleport/blob/74591e4ad1a91764b2976a8b1066c3635d3e9dd7/lib/config/openssh/openssh.go#L48-L56

So I don't think we need to change anything here.

or maybe i misunderstood the question that you are referring to something else?